File: semester1/Test3/num1/list.cpp
#include <iostream>
#include "list.h"
using namespace std;

List *createList()
{
    return new List {nullptr};
}

void deleteList(List *list)
{
    ListElement *current = list->first;
    while (current)
    {
        list->first = current->next;
        delete current;
        current = list->first;
    }

    delete list;
}

void addElement(List *list, int number)
{
    if (isEmpty(list))
    {
        list->first = new ListElement {number, 1, list->first};
        return;
    }

    ListElement *current = list->first;

    if (exists(list, number))
    {
        while (current->number != number)
        {
            current = current->next;
        }

        current->amount++;

        return;
    }

    list->first = new ListElement {number, 1, list->first};

    while (current->next && current->number > current->next->number)
    {
        swap(current->amount, current->next->amount);
        swap(current->number, current->next->number);
        current = current->next;
    }
}

void removeElement(ListElement *&listElement, int number)
{
    if (!listElement)
    {
        return;
    }

    if (listElement->number == number)
    {
        ListElement *extra = listElement->next;
        delete listElement;
        listElement = extra;
        return;
    }

    removeElement(listElement->next, number);
}

void removeElement(List *list, int number)
{
    removeElement(list->first, number);
}

void printList(List *list)
{
    ListElement *current = list->first;
    while (current)
    {
        cout << current->number << ": " << current->amount << " times" << endl;
        current = current->next;
    }
}

bool isEmpty(List *list)
{
    return !(list->first);
}

bool exists(List *list, int number)
{
    if (isEmpty(list))
    {
        return false;
    }

    ListElement *current = list->first;

    while (current)
    {
        if (current->number == number)
        {
            return true;
        }

        current = current->next;
    }

    return false;
}

File: semester2/hw4/num1/src/main/java/com/artemchernikov/g144/List.java
package com.artemchernikov.g144;

/**A template class describing singly linked list*/
public class List<T> {

    private ListElement first;
    private int size;

    /**
     * A method adds element to the head of the list
     * @param value element to add
     * */
    public void addElement(T value) {
        first = new ListElement(value, first);
        this.size++;
    }

    /**
     * A method removes element from the list, if the list contains it
     * @param value element to remove
     * */
    public void removeElement(T value) {
        if (!isEmpty()) {
            if (value.equals(first.value)) {
                first = first.next;
                this.size--;
                return;
            }

            ListElement previous = first;
            ListElement current = first.next;

            while (current != null) {
                if (value.equals(current.value)) {
                    previous.next = current.next;
                    this.size--;
                    return;
                }

                previous = current;
                current = current.next;
            }
        }
    }

    /**
     * A method returns value by index
     * @param index index of needed element
     * @throws IndexOutOfBoundsException if received index is out of bounds
     * */
    public T getElement(int index) throws IndexOutOfBoundsException {
        if (index >= 0 && index < getSize()) {
            ListElement current = first;
            for (int i = 0; i < index; i++) {
                current = current.next;
            }

            return current.value;
        } else {
            throw new IndexOutOfBoundsException();
        }
    }

    /**
     * A method checks if element exists in the list
     * @param value element to check
     * @return true if the list contains received value and false otherwise
     * */
    public boolean exists(T value) {
        ListElement current = first;
        while (current != null) {
            if (value.equals(current.value)) {
                return true;
            }
            current = current.next;
        }

        return false;
    }

    /**
     * A method returns number of list elements
     * @return number of list elements
     * */
    public int getSize() {
        return this.size;
    }

    /**
     * A method checks if the list is empty
     * @return true if the list is empty and false otherwise
     * */
    public boolean isEmpty() {
        return first == null;
    }

    /**A method displays the list*/
    public void print() {
        ListElement current = first;

        while (current != null) {
            System.out.println(current.value);
            current = current.next;
        }
    }

    /**A class describing element of singly linked list*/
    private class ListElement {
        private T value;
        private ListElement next;

        private ListElement(T value, ListElement next) {
            this.value = value;
            this.next = next;
        }
    }
}
